File: Assignment2/c/palindrome.cpp
#include <iostream>
#include <string>

#include "queue.h"
#include "stack.h"

inline bool is_palindrome_stack(std::string input) {
    Stack<char> s =  Stack<char>();
    Queue<char> q =  Queue<char>();
    for (char ch : input) {
        s.push(ch);
        q.enqueue(ch);
    }
    while (!s.isEmpty() && !q.isEmpty())
        if (s.pop() != q.dequeue())
            return false;
    return true;
}

inline int p_main() {
    std::cout << std::boolalpha;
    std::cout <<"\"racecar\"\t"<< (is_palindrome_stack("racecar")) << std::endl;  // true
    std::cout <<"\"x\"\t\t\t"<< (is_palindrome_stack("x")) << std::endl;        // one character
    std::cout <<"\"\"\t\t\t"<< (is_palindrome_stack("")) << std::endl;         // null case
    std::cout <<"\"sample\"\t"<< (is_palindrome_stack("sample")) << std::endl;     // false
    std::cout <<"\"dab\"\t\t"<< (is_palindrome_stack("dab")) << std::endl;       // false
    std::cout <<"\"<>\"\t\t"<< (is_palindrome_stack("<>")) << std::endl;       // false
    return 0;
}

File: Assignment2/j/Palindrome.java
package Assignment2.j;

public class Palindrome {
    public static boolean is_palindrome(String input) {
        Stack<Character> s = new Stack<>();
        Queue<Character> q = new Queue<>();
        for (char ch : input.toCharArray()) {
            s.push(ch);
            q.enqueue(ch);
        }
        while (!s.isEmpty() && !q.isEmpty())
            if (s.pop() != q.dequeue())
                return false;
        return true;
    }

    public static void p_main() {
        System.out.println("\"racecar\"\t" + is_palindrome("racecar"));
        System.out.println("\"x\"\t\t\t" + is_palindrome("x"));
        System.out.println("\"\"\t\t\t" + is_palindrome(""));
        System.out.println("\"sample\"\t" + is_palindrome("sample"));
        System.out.println("\"dab\"\t\t" + is_palindrome("dab"));
        System.out.println("\"<>\"\t\t" + is_palindrome("<>"));
    }
}
